# CS109A Project Group 21
# Baseline Model

*****

In [12]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import NearestNeighbors
from scipy.sparse import csr_matrix
from collections import Counter

### Importing the relevant datasets

First we import two subset of 100 playlists, each subset of 100 playlists to be used for the training and test sets respectively. 

These playlists are randomly chosen from the full one million playlist dataset that has been matched with the Spotify API data, thus including data on track-specific features such as loudness, speechiness and tempo.

In [14]:
train_df = pd.read_csv("data/subset100playlists.csv")
test_df = pd.read_csv("data/subset100playlists_test.csv")

In [15]:
train_df.head()

Unnamed: 0  index   pid  pos                             track_uri  \
0           0    107  6011    0  spotify:track:2FBUoWkIuQXwayw2RNo5l6   
1           1    107  6011    1  spotify:track:25khomWgBVamSdKw7hzm3l   
2           2    107  6011    2  spotify:track:2NVt7fxr5GsqTkGwYXcNTE   
3           3    107  6011    3  spotify:track:5NQbUaeTEOGdD6hHcre0dZ   
4           4    107  6011    4  spotify:track:5hJFhO9dvhJoDvUZZ9iWSw   

                            artist_uri_x             track_name  \
0  spotify:artist:3TVXtAsR1Inumwj472S9r4          Hotline Bling   
1  spotify:artist:1Xyo4u8uXC1ZmMpatF05PJ              The Hills   
2  spotify:artist:5lHRUCqkQZCIWeX7xG4sYT   Flex (Ooh, Ooh, Ooh)   
3  spotify:artist:6PXS4YHDkKvl1wkIl4V8DL  679 (feat. Remy Boyz)   
4  spotify:artist:360IAlyVv4PCEVjgyMZrxK                  Adorn   

       artist_name                           artist_uri_y  \
0            Drake  spotify:artist:3TVXtAsR1Inumwj472S9r4   
1       The Weeknd  spotify:artist:1Xyo4u8uXC1ZmMpatF05PJ   
2  Rich Homie Quan  spotify:artist:5lHRUCqkQZCIWeX7xG4sYT   
3        Fetty Wap  spotify:artist:6PXS4YHDkKvl1wkIl4V8DL   
4           Miguel  spotify:artist:360IAlyVv4PCEVjgyMZrxK   

                            album_name  ... key  liveness  loudness mode  \
0                                Views  ...   2    0.0504    -8.094    1   
1            Beauty Behind The Madness  ...   0    0.1370    -7.011    0   
2                 Flex (Ooh, Ooh, Ooh)  ...   6    0.3410    -7.534    0   
3                679 (feat. Remy Boyz)  ...   7    0.6250    -5.738    1   
4  Kaleidoscope Dream (Deluxe Version)  ...  11    0.1870    -5.693    0   

   speechiness    tempo  time_signature  \
0       0.0587  134.960               4   
1       0.0798  135.553               5   
2       0.0496  163.993               4   
3       0.3180  190.050               4   
4       0.1750  179.063               4   

                                          track_href            type  valence  
0  https://api.spotify.com/v1/tracks/2FBUoWkIuQXw...  audio_features    0.539  
1  https://api.spotify.com/v1/tracks/25khomWgBVam...  audio_features    0.109  
2  https://api.spotify.com/v1/tracks/2NVt7fxr5Gsq...  audio_features    0.883  
3  https://api.spotify.com/v1/tracks/5NQbUaeTEOGd...  audio_features    0.603  
4  https://api.spotify.com/v1/tracks/5hJFhO9dvhJo...  audio_features    0.235  

[5 rows x 29 columns]

### Creating the train and test set

We then create the train and test set by keeping select numeric features and the track URI, as a way to maintain a unique ID for each song.

Specifically, for the test set, we split it into two subsets: calibration (20%) and withheld songs (80%), where the calibration songs are used as inputs to the model to predict songs. These predicted songs are then compared to the withheld songs as a comparison metric.

In [18]:
numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
df_cleaned= train_df.select_dtypes(include=numerics)

var_drop = ["index","pid","pos", "count", "Unnamed: 0"]
df_cleaned = df_cleaned.drop(var_drop, axis =1)

train = pd.concat([df_cleaned, train_df['track_uri']],axis=1)

In [19]:
test_df_cleaned = test_df.select_dtypes(include=numerics)

test_var_drop = ["index","pos", "count", "Unnamed: 0"]
test_df_cleaned = test_df_cleaned.drop(test_var_drop, axis =1)

test_df_cleaned = pd.concat([test_df_cleaned, test_df['track_uri']],axis=1)
calibration, withheld = train_test_split(test_df_cleaned, test_size=0.8, random_state=209, stratify = test_df_cleaned['pid'])

In [20]:
train.describe()

acousticness  danceability    duration_ms        energy  \
count  13562.000000  13562.000000   13562.000000  13562.000000   
mean       0.157796      0.651324  234039.366023      0.685388   
std        0.198003      0.148380   51493.366351      0.167506   
min        0.000006      0.144000   19133.000000      0.027900   
25%        0.020000      0.549000  202723.250000      0.574000   
50%        0.073450      0.658000  225991.500000      0.702000   
75%        0.222000      0.757000  256703.500000      0.813000   
max        0.987000      0.980000  826813.000000      0.996000   

       instrumentalness           key      liveness      loudness  \
count      13562.000000  13562.000000  13562.000000  13562.000000   
mean           0.012731      5.128300      0.186206     -6.434418   
std            0.076586      3.668483      0.144203      2.653691   
min            0.000000      0.000000      0.015700    -34.475000   
25%            0.000000      1.000000      0.092000     -7.703000   
50%            0.000000      5.000000      0.128000     -5.907000   
75%            0.000048      8.000000      0.249000     -4.574750   
max            0.987000     11.000000      1.000000     -0.698000   

               mode   speechiness         tempo  time_signature       valence  
count  13562.000000  13562.000000  13562.000000    13562.000000  13562.000000  
mean       0.647028      0.110930    122.022105        3.978985      0.553528  
std        0.477912      0.109549     28.255147        0.260324      0.226203  
min        0.000000      0.022500     34.717000        0.000000      0.034000  
25%        0.000000      0.038500     99.973250        4.000000      0.382000  
50%        1.000000      0.060500    120.076000        4.000000      0.559000  
75%        1.000000      0.141000    140.000000        4.000000      0.737750  
max        1.000000      0.956000    210.750000        5.000000      0.990000

### Dropping duplicates
In the baseline model, given that we are only looking at predicting songs based on their similarity to other tracks, we can drop duplicates of the same song from the training set. We do so below.

In [21]:
train.drop_duplicates(subset ="track_uri",keep = False, inplace = True) 
train.head()

acousticness  danceability  duration_ms  energy  instrumentalness  key  \
5        0.00563         0.800       159693   0.677           0.00409    0   
12       0.02940         0.758       222180   0.756           0.00000   10   
19       0.19600         0.855       238530   0.522           0.00000    3   
22       0.10700         0.703       214757   0.806           0.00000    2   
26       0.57700         0.670       130293   0.724           0.00000   11   

    liveness  loudness  mode  speechiness    tempo  time_signature  valence  \
5     0.0838    -4.023     1       0.0393  125.004               4    0.547   
12    0.0906    -5.059     0       0.1380  145.068               4    0.406   
19    0.1000    -5.540     1       0.2260  139.917               4    0.355   
22    0.1340    -6.415     1       0.1530   94.979               4    0.279   
26    0.7890    -4.141     1       0.3610  145.105               4    0.285   

                               track_uri  
5   spotify:track:5bcfWAHLoMfGBdbeyoMaH3  
12  spotify:track:1rZah1UGXvgiQCqLVlq4hS  
19  spotify:track:2okC0OWtgS86P2LC18IzHy  
22  spotify:track:2FdQ4hkbqZ1X930oxxgZZy  
26  spotify:track:1Wsbr1d2BouNGk2q92mIj7

In [22]:
train = train.set_index('track_uri') 

In [23]:
train.shape

(3098, 13)

### Fitting the training set to a Nearest Neighbors model

This leaves us with 3098 songs in our training set, which we then fit onto an unsupervised Nearest Neighbors learning model, where n_neighbors = 20 and cosine distance is used as a distance of measure. 

Here, we have set the neighbors search algorithm to auto, so that the algorithm attempts to determine the best approach from the training data across the options of BallTree, KDTree, and a brute-force algorithm.

In [25]:
# pivot ratings into song features
model_knn = NearestNeighbors(metric='cosine', algorithm='auto', n_neighbors=20, n_jobs=-1)# fit the dataset
model_knn.fit(train)

NearestNeighbors(algorithm='auto', leaf_size=30, metric='cosine',
         metric_params=None, n_jobs=-1, n_neighbors=20, p=2, radius=1.0)

### Creating a function to make recommendations

We then proceed to create a function that will take as inputs 1) the Nearest Neighbors model fitted on the training data, 2) a playlist id (to identify the songs in the test set), 3) the number of neighbors to select from and 4) number of recommended songs to predict.

This function will return a list of recommended songs.

In [26]:
def make_recommendation(model_knn, playlist_id, n_neighbors, n_recommendations):
    
    calibration_songs = calibration[calibration['pid']==playlist_id]
    calibration_songs_clean = calibration_songs.drop(columns=['pid', 'track_uri'])

    song_freq_dict = {}

    for index, song in calibration_songs_clean.iterrows():
        distances, indices = model_knn.kneighbors(song.values.reshape(1,-1), n_neighbors=n_neighbors)

        for index in indices[0]:
            if song_freq_dict.get(index) is None:
                song_freq_dict[index] = 1
            else:
                song_freq_dict[index] += 1

    k = Counter(song_freq_dict) 

    # Finding n highest values 
    top_songs = [i[0] for i in k.most_common(n_recommendations)]    

    rec_songs = train.iloc[top_songs].index

    return rec_songs

### Measuring success

For starters, we will use the R-precision score to measure how successful our song predictions are. Here, we define a function `r_precision`, which computes number of retrieved relevant tracks divided by the number of known relevant tracks (i.e., the number of withheld tracks). It takes as inputs a list of predicted songs and compares it against a list of the actual songs from the given playlist ("withheld songs").

In [27]:
#Function to calculate r_precision (https://recsys-challenge.spotify.com/rules)
#R-precision is the number of retrieved relevant tracks divided by the number of known relevant tracks (i.e., the number of withheld tracks) 

def r_precision(preds, known):
    for i in known:
        if i in preds:
            song = combined_df[combined_df['track_uri']==i]['track_name']
            print(f'{song} appeared in both our predicted playlist and the known list of songs.')
    score = np.sum(known.isin(preds))/known.shape[0]
    return score

### Running the baseline model on test data

We then run the following code to loop through all the unique playlists in the test set to predict songs based on a random subset of 20% of songs from the given test playlist (`calibration`). The predicted songs are then fed into the R-precision score function together with the remaining 80% of songs from the given test playlist (`withheld`).

For testing purposes, we print out the actual track names of the songs being predicted and the withheld songs, so that we can get a qualitative understanding of the songs.

In [29]:
n_neighbors, n_recommendations = 50, 100
r_precision_scores = []

#Loop through all unique playlists in test set to identify predicted songs
for index, pid in enumerate(withheld['pid'].drop_duplicates()): 
    print (index, pid)
    pred_songs = make_recommendation(model_knn, pid, n_neighbors, n_recommendations)
    validation_set = withheld[withheld.pid == pid].track_uri
    print("Predicted songs\n", combined_df[combined_df['track_uri'].isin(pred_songs)]['track_name'])
    print("Known songs\n", test_df[test_df['track_uri'].isin(validation_set)]['track_name'])
    
    rps = r_precision(list(pred_songs), validation_set)
    r_precision_scores.append(rps)
    
    print(f'Playlist {pid}: The R precision score is {rps}')

0 198457
Predicted songs
 66                              Wat U Mean (Aye, Aye, Aye)
234                                          No Letting Go
300                                              R U Mine?
350                                             Obstacle 1
677                                                   Donk
838                                          Body Language
842                                         Si Tú No Estás
1045                             Bright Lights Bigger City
1267                                Let It Go - feat. Akon
1340     Feeling Right (Everything Is Nice) [feat. Popc...
1384                                         Broken Arrows
1405                      She Doesn't Mind - Pitbull Remix
1417                       Send It (feat. Rich Homie Quan)
1461                                      One In A Million
1713                                      Let Me Down Easy
1740                                       Til My Last Day
1866                          

Predicted songs
 214                                                Ice Box
234                                          No Letting Go
485            Chickenhead (feat. La Chat & Three 6 Mafia)
763                                                 Hollow
791                                           Pretty Hurts
1232                                              No Limit
1375                        The Nights - Felix Jaehn Remix
1551                    Strut - From "The Cheetah Girls 2"
1792                                     That Should Be Me
1900                You Can't Stop Me (feat. Thomas Rhett)
2157                                                  Love
2159                                   Somewhere In My Car
2517                                              Everlong
2869                                            Break Free
2878                                            Ugly Heart
3103             Should I Stay or Should I Go - Remastered
3418                                   

Predicted songs
 76                                              Came To Do
144                               Primetime (feat. Miguel)
300                                              R U Mine?
339                                    Steady, As She Goes
692                                 Wop (Official Version)
838                                          Body Language
1251     Habits (Stay High) - The Chainsmokers Extended...
1320                                     Should’ve Been Us
1321                                    Once In A Lifetime
1342                               El Perdón - Forgiveness
1387                                                  Zimt
1392                                Magnets - A-Trak Remix
1531                                         Dance With Me
1674                                                  Amen
1691                                                   Fly
1740                                       Til My Last Day
1877                                   

Predicted songs
 145                         LoveHate Thing - feat. Sam Dew
334                                             Blue Jeans
480                                            Like A Pimp
517                                                  Whoa!
675                                         Feelin' Myself
910      Let's Go (feat. Big D & Twista) - explicit alb...
1232                                              No Limit
1493                                           Stacy's Mom
1807                                              Been You
1900                You Can't Stop Me (feat. Thomas Rhett)
2621                                    Fell On Black Days
2829                                      Waiting For Love
3187                                             Nightrain
3718                                               Tell Em
4015                                          Last Request
4034              Trap Paris (feat. Quavo & Ty Dolla $ign)
4064                              Frien

Predicted songs
 88                      You Don't Own Me - Candyland Remix
334                                             Blue Jeans
492                                   Sippin On Some Syrup
761                                                Perfect
1056                               Stanky Legg - Main Edit
1089                                           Trap N*ggas
1209                                  Ridin' dirty (Remix)
1222                          We Be Burnin' (Recognize It)
1322                                     Let's Do It Again
1406     This One's For You (feat. Zara Larsson) - Offi...
1454                                                 90210
1692                              Ladies Love Country Boys
1713                                      Let Me Down Easy
1839                              Every Rose Has Its Thorn
1906                                 Tell Me Where To Park
1981                                          Time Is Love
2074                                   

Predicted songs
 45                                                     WOA
175                                                Foolish
320                                                 Domo23
570                                         Can I Kick It?
571                                             C.R.E.A.M.
842                                         Si Tú No Estás
1367                                    Whole Lotta Lovin'
1417                       Send It (feat. Rich Homie Quan)
1599                                              Good Man
1682                             Never Wanted Nothing More
1892                                           Kiss a Girl
1904                                   If You Were My Girl
1967                                        Goodnight Kiss
2055                                                 Banjo
2058                                            We Were Us
2158                                 Kick It In The Sticks
2374                                   

Predicted songs
 523                                         Love Like This
584                  Return Of The Mack - C & J Street Mix
715                                                 Switch
720                                              Beat Down
951                                               Bend Ova
1393                                                   Sex
1454                                                 90210
1599                                              Good Man
1865                                              Be Still
2046                                    Somewhere With You
2174     Long Cool Woman (In A Black Dress) - 1999 Rema...
2300                        Where The Streets Have No Name
2481                           Waiting For A Girl Like You
2605                                            No Excuses
2634                                               Seether
2703                                        Scream & Shout
2822                                   

Predicted songs
 79                                            Gucci Snakes
117                                            sobeautiful
233                                Slow Down - 12" Version
242                                                Promise
530                                  Thuggish Ruggish Bone
587                                         Patron Tequila
795          Beautiful Liar - Main Version / Album Version
877                                                   Army
1027     All The Above [feat. T-Pain] - Explicit Album ...
1340     Feeling Right (Everything Is Nice) [feat. Popc...
1384                                         Broken Arrows
1432                       The Water Dance (feat. Pitbull)
1508                                        Start All Over
1509                                            I Miss You
1792                                     That Should Be Me
1880                                             Apartment
1973                                   

Predicted songs
 117                                sobeautiful
144                   Primetime (feat. Miguel)
150                                Rich Niggaz
165                     Neighbors Know My Name
369                     Fluorescent Adolescent
738           Home - Fedde Le Grand Radio Edit
871                                Rock Bottom
1367                        Whole Lotta Lovin'
1534                       She's Like The Wind
1592      19 Somethin' - Greatest Hits Version
1892                               Kiss a Girl
1993                                     Gonna
2058                                We Were Us
2125                                      This
2747                             Wobble Wobble
2866     Prayer in C - Robin Schulz Radio Edit
2878                                Ugly Heart
2924                        What Would You Say
3126                                 Don't Cry
3192                          Behind Blue Eyes
3337                               4r Da Sq

Predicted songs
 22                                               Big Tymin
82                                      Look Alive - Remix
125                                               Novacane
146                                           Made to Love
175                                                Foolish
246                                                 Get Up
341                                          Ultraviolence
479                                                     Ha
497                        Can't Deny It - feat. Nate Dogg
570                                         Can I Kick It?
584                  Return Of The Mack - C & J Street Mix
650                                       Get Buck In Here
720                                              Beat Down
767                                  Writing's On The Wall
1045                             Bright Lights Bigger City
1253                                         Happy Endings
1311                            Another

Predicted songs
 326                                               I. crawl
423                                            Let's Dance
504                                          Passin' Me By
513                     Damn It Feels Good to Be a Gangsta
570                                         Can I Kick It?
719                                    Levels - Radio Edit
1193     I Think They Like Me - feat. Jermaine Dupri, D...
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1733                                            My Old Man
2198                                             Strangers
2202                                     Piece of My Heart
2300                        Where The Streets Have No Name
2328                            Bridge Over Troubled Water
2905       2 Minutes To Midnight - 1998 Remastered Version
2968                               Working for the Weekend
3116                                          School's Out
3140                                   

Predicted songs
 79                                            Gucci Snakes
242                                                Promise
338                                        Make It Wit Chu
481                                             nann nigga
485            Chickenhead (feat. La Chat & Three 6 Mafia)
530                                  Thuggish Ruggish Bone
795          Beautiful Liar - Main Version / Album Version
1156                                               Lucifer
1273                      Church - Main Version - Explicit
1288                Five More Hours - Deorro x Chris Brown
1302                                  Heroes (we could be)
1456                           Rockstar - Original Version
1559     Only You-Bad Boy Remix (feat. The Notorious B....
1687                                      Shut Up And Fish
1792                                     That Should Be Me
1808                                        Get Used To It
1973                                   

Predicted songs
 252                                       Shake It Off
303                                         In One Ear
721                                     Air Force Ones
1250                        Save the World - Radio Mix
1303                                          Geronimo
1358                                           Save Me
1457                                  We Got The Party
1596                                            Mayday
1688                                   Write This Down
2091                                 Mine Would Be You
2133                                     Jump Right In
2136                                 Bumpin' The Night
2140                                        Neon Light
2310         Learning To Fly - 2011 Remastered Version
2430                          Little Lies - Remastered
2511           Hey Man, Nice Shot - Remastered Version
2814                                            Shower
3209                                    Black Ho

Predicted songs
 117                                            sobeautiful
124                  Boyfriend #2 - Explicit Album Version
144                               Primetime (feat. Miguel)
150                                            Rich Niggaz
423                                            Let's Dance
719                                    Levels - Radio Edit
842                                         Si Tú No Estás
1296                                     Can't Trust Thots
1334                           Nobody To Love - Radio Edit
1387                                                  Zimt
1405                      She Doesn't Mind - Pitbull Remix
1415                          Let Me Hold You (Turn Me On)
1416                                     Perfect Strangers
1498                                Cleanin' Out My Closet
1592                  19 Somethin' - Greatest Hits Version
1674                                                  Amen
1681                                   

Predicted songs
 112                                       Diced Pineapples
334                                             Blue Jeans
480                                            Like A Pimp
493                                    Ruff Ryders' Anthem
537                                                 I Wish
729                                         Nobody To Love
825                                            Don't Worry
1464                                           She Wildin'
1688                                       Write This Down
1708                                   Real Men Love Jesus
1900                You Can't Stop Me (feat. Thomas Rhett)
2043                              Here's To The Good Times
2055                                                 Banjo
2121                                               Alright
2332                                          Already Gone
2374                                          Have a Cigar
2511               Hey Man, Nice Shot -

Predicted songs
 61                                 Campaign (feat. Future)
98                            Pull Up (feat. Lil Uzi Vert)
137                                             Love Faces
206                                     Aston Martin Music
728                                    Peanut Butter Jelly
738                       Home - Fedde Le Grand Radio Edit
842                                         Si Tú No Estás
1307                                            The Illest
1360                                                  Mary
1384                                         Broken Arrows
1482               Dirt Off Your Shoulder / Lying From You
2130                                  Small Town Throwdown
2157                                                  Love
2159                                   Somewhere In My Car
2548                                        Ain't My Fault
2583                                                Bodies
2608                                   

Predicted songs
 125                                               Novacane
202                                           Shake Ya Ass
504                                          Passin' Me By
570                                         Can I Kick It?
767                                  Writing's On The Wall
840                                          Lejos De Aquí
1131                                               Mercy.1
1193     I Think They Like Me - feat. Jermaine Dupri, D...
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1273                      Church - Main Version - Explicit
1903                                       Time Well Spent
2055                                                 Banjo
2300                        Where The Streets Have No Name
2455                   I Want To Break Free - Single Remix
2518                                        I Will Survive
2601                                      Freak On a Leash
2644                                   

Predicted songs
 28                                                Feedback
150                                            Rich Niggaz
214                                                Ice Box
482                       Big Booty Hoes (feat. Too $hort)
525                                         Me, Myself & I
1446                           Don't Leave - Gryffin Remix
1692                              Ladies Love Country Boys
1873                                        Guns For Hands
2545                                            in my miNd
2605                                            No Excuses
2998                                             The Metal
3412                                           Gyalchester
3456                                               What If
3695                                              Fireball
3831             Slow Jamz (feat. Kanye West & Jamie Foxx)
4035                                            Loving You
4205                                   

Predicted songs
 107                                              Do You...
214                                                Ice Box
233                                Slow Down - 12" Version
313                                                My Type
339                                    Steady, As She Goes
480                                            Like A Pimp
581                 What's Your Fantasy (Featuring Shawna)
825                                            Don't Worry
877                                                   Army
1285     Crazy Youngsters - From "Pitch Perfect 2" Soun...
1340     Feeling Right (Everything Is Nice) [feat. Popc...
1384                                         Broken Arrows
1404                                 Heatwave (feat. Akon)
1405                      She Doesn't Mind - Pitbull Remix
1498                                Cleanin' Out My Closet
1508                                        Start All Over
1509                                   

Predicted songs
 76                                              Came To Do
331                           Run [I'm A Natural Disaster]
734                                    Move (If You Wanna)
1156                                               Lucifer
1203                                All The Way Up (Remix)
1334                           Nobody To Love - Radio Edit
1534                                   She's Like The Wind
1596                                                Mayday
1699                                         Alone Tonight
1986                                      Kiss You Tonight
2829                                      Waiting For Love
2990                         Rio - 2009 Remastered Version
3070                                   Rock Of Ages (2012)
3193                                            Yesterdays
3196                                      One In A Million
3342                                          Digital Love
3415                                   

Predicted songs
 329                                         Run the Jewels
537                                                 I Wish
541                    U.N.I.T.Y. - Album Version (Edited)
675                                         Feelin' Myself
737                     Pumpin Blood - The Jane Doze Remix
743               We Own The Night - The Chainsmokers Edit
886                                        Let Me Love You
1084                                            Basketball
1393                                                   Sex
1986                                      Kiss You Tonight
2060                                    You'll Think Of Me
2300                        Where The Streets Have No Name
2337                                      The Longest Time
2621                                    Fell On Black Days
2829                                      Waiting For Love
3113                                 Mama, I'm Coming Home
3255                                   

Predicted songs
 26                                                   Pt. 2
28                                                Feedback
118                                     Break From Toronto
485            Chickenhead (feat. La Chat & Three 6 Mafia)
497                        Can't Deny It - feat. Nate Dogg
504                                          Passin' Me By
926                             El Taxi - Original Version
1104                                              Research
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1375                        The Nights - Felix Jaehn Remix
3038                                          D'yer Mak'er
3337                                           4r Da Squaw
3345                                           Nightvision
3467                                       Life Ain't Easy
3482                                   U-RITE - Rynx Remix
3547                                         Rockin' Robin
3805                                   

Predicted songs
 5                              Intoxicated - New Radio Mix
498                                              Xxplosive
671      Love Me Like You Do - From "Fifty Shades Of Grey"
749                                      Bitch I'm Madonna
877                                                   Army
1093                                            RGF Island
1212                                            Move Bitch
1464                                           She Wildin'
1541                                               We Rock
1602                                        Next Boyfriend
1708                                   Real Men Love Jesus
1904                                   If You Were My Girl
1924                                             Ex To See
2055                                                 Banjo
2076                                                  Dust
2128                                                Voices
2408                                   

Predicted songs
 19                          Bake Sale (feat. Travis Scott)
76                                              Came To Do
79                                            Gucci Snakes
117                                            sobeautiful
242                                                Promise
395                                           Whip My Hair
529                                         Knockin' Boots
530                                  Thuggish Ruggish Bone
587                                         Patron Tequila
734                                    Move (If You Wanna)
795          Beautiful Liar - Main Version / Album Version
812                                       Some Type of Way
1027     All The Above [feat. T-Pain] - Explicit Album ...
1156                                               Lucifer
1251     Habits (Stay High) - The Chainsmokers Extended...
1321                                    Once In A Lifetime
1342                               El P

Predicted songs
 44                                          Oh My Dis Side
104                                               Doorbell
219                           I Invented Sex [feat. Drake]
338                                        Make It Wit Chu
347                              Dangerous (feat. Joywave)
353                                               Lazy Eye
479                                                     Ha
497                        Can't Deny It - feat. Nate Dogg
523                                         Love Like This
716                                              Got Money
926                             El Taxi - Original Version
1045                             Bright Lights Bigger City
1312                      Prayer In C - Robin Schulz Remix
1351                                   Fun (feat. Tove Lo)
1559     Only You-Bad Boy Remix (feat. The Notorious B....
1772     Blank Space/Stand By Me - Live From Spotify Lo...
1793                       Runaway Love

Predicted songs
 117                                       sobeautiful
148                                           Fantasy
150                                       Rich Niggaz
592                                Often - Kygo Remix
655                                       Caribou Lou
675                                    Feelin' Myself
749                                 Bitch I'm Madonna
1212                                       Move Bitch
1322                                Let's Do It Again
1424                                            Alarm
1464                                      She Wildin'
1592             19 Somethin' - Greatest Hits Version
1688                                  Write This Down
1708                              Real Men Love Jesus
1904                              If You Were My Girl
2055                                            Banjo
2125                                             This
2326                      Over The Hills And Far Away
2372       

Predicted songs
 22                                               Big Tymin
125                                               Novacane
165                                 Neighbors Know My Name
175                                                Foolish
326                                               I. crawl
369                                 Fluorescent Adolescent
570                                         Can I Kick It?
738                       Home - Fedde Le Grand Radio Edit
767                                  Writing's On The Wall
1089                                           Trap N*ggas
1312                      Prayer In C - Robin Schulz Remix
1371                                          Stick Around
1708                                   Real Men Love Jesus
1904                                   If You Were My Girl
2055                                                 Banjo
2374                                          Have a Cigar
2518                                   

Predicted songs
 125                                               Novacane
300                                              R U Mine?
350                                             Obstacle 1
738                       Home - Fedde Le Grand Radio Edit
1104                                              Research
1193     I Think They Like Me - feat. Jermaine Dupri, D...
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1320                                     Should’ve Been Us
1392                                Magnets - A-Trak Remix
1493                                           Stacy's Mom
1498                                Cleanin' Out My Closet
1743                                        Temporary Home
1793                       Runaway Love - Kanye West Remix
2130                                  Small Town Throwdown
2157                                                  Love
2159                                   Somewhere In My Car
2514                                  B

Predicted songs
 125                                               Novacane
233                                Slow Down - 12" Version
479                                                     Ha
715                                                 Switch
841                                                   6 AM
877                                                   Army
926                             El Taxi - Original Version
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1253                                         Happy Endings
1389         I'd Love To Change The World - Matstubs Remix
1508                                        Start All Over
1509                                            I Miss You
1691                                                   Fly
2618                                           Just A Girl
2822                                           Locked Away
3019                                           Hells Bells
3125                                   

Predicted songs
 117                                            sobeautiful
135                                        Wu-Tang Forever
150                                            Rich Niggaz
511      Put Your Hands Where My Eyes Could See - Expli...
715                                                 Switch
723             Habits (Stay High) - Hippie Sabotage Remix
1551                    Strut - From "The Cheetah Girls 2"
1579                                           Bait A Hook
1685                                                  Stay
1734                               I Could Use a Love Song
1786                                       Yellow Raincoat
1808                                        Get Used To It
2313                                             Limelight
2339                             The Ballad Of Curtis Loew
3060                                        Mr. Brownstone
3100                        Bicycle Race - Remastered 2011
3274                                   

Predicted songs
 116                                               Falsetto
385                                      Keep Your Head Up
423                                            Let's Dance
482                       Big Booty Hoes (feat. Too $hort)
791                                           Pretty Hurts
881                    Mind Your Manners (feat. Icona Pop)
948                                          Calabria 2008
1232                                              No Limit
1262         Bout Me - feat. Problem & Iamsu [Bonus Track]
1271                              Jealous - Bakermat Remix
1290                            Everything Is Embarrassing
1296                                     Can't Trust Thots
1866                                                 Human
1873                                        Guns For Hands
1880                                             Apartment
1900                You Can't Stop Me (feat. Thomas Rhett)
2297                                   

Predicted songs
 104                                               Doorbell
155                                                So Gone
165                                 Neighbors Know My Name
214                                                Ice Box
485            Chickenhead (feat. La Chat & Three 6 Mafia)
495                    Int'l Players Anthem (I Choose You)
525                                         Me, Myself & I
570                                         Can I Kick It?
716                                              Got Money
791                                           Pretty Hurts
1303                                              Geronimo
1375                        The Nights - Felix Jaehn Remix
1551                    Strut - From "The Cheetah Girls 2"
1735                                     Five More Minutes
1772     Blank Space/Stand By Me - Live From Spotify Lo...
1901                                          Lose My Mind
1904                                   

Predicted songs
 137                                             Love Faces
260                                             I Want Her
368                                              Lazaretto
485            Chickenhead (feat. La Chat & Three 6 Mafia)
737                     Pumpin Blood - The Jane Doze Remix
825                                            Don't Worry
948                                          Calabria 2008
1084                                            Basketball
1367                                    Whole Lotta Lovin'
1375                        The Nights - Felix Jaehn Remix
1792                                     That Should Be Me
1892                                           Kiss a Girl
1900                You Can't Stop Me (feat. Thomas Rhett)
1981                                          Time Is Love
2058                                            We Were Us
2159                                   Somewhere In My Car
2297                                   

Predicted songs
 88                      You Don't Own Me - Candyland Remix
117                                            sobeautiful
219                           I Invented Sex [feat. Drake]
242                                                Promise
258                                            Sex Therapy
347                              Dangerous (feat. Joywave)
349                                                    PDA
492                                   Sippin On Some Syrup
523                                         Love Like This
539                                          Ms. Fat Booty
587                                         Patron Tequila
840                                          Lejos De Aquí
1191     I Know You See It (feat. Brandy "Ms. B" Hambri...
1322                                     Let's Do It Again
1406     This One's For You (feat. Zara Larsson) - Offi...
1793                       Runaway Love - Kanye West Remix
1973                                   

Predicted songs
 88                      You Don't Own Me - Candyland Remix
111                                                  waves
326                                               I. crawl
492                                   Sippin On Some Syrup
808                                                  Ginza
948                                          Calabria 2008
1084                                            Basketball
1209                                  Ridin' dirty (Remix)
1232                                              No Limit
1375                        The Nights - Felix Jaehn Remix
1406     This One's For You (feat. Zara Larsson) - Offi...
1454                                                 90210
1692                              Ladies Love Country Boys
1866                                                 Human
1881                                               My Body
2055                                                 Banjo
2212                    Suite: Judy Blu

Predicted songs
 22                                               Big Tymin
66                              Wat U Mean (Aye, Aye, Aye)
146                                           Made to Love
584                  Return Of The Mack - C & J Street Mix
671      Love Me Like You Do - From "Fifty Shades Of Grey"
720                                              Beat Down
732                                             Girlfriend
881                    Mind Your Manners (feat. Icona Pop)
1290                            Everything Is Embarrassing
1301                        Bad (feat. Vassy) - Radio Edit
1393                                                   Sex
1461                                      One In A Million
1541                                               We Rock
1587                                                  Then
1924                                             Ex To See
2046                                    Somewhere With You
2469                             I Was 

Predicted songs
 76                                              Came To Do
79                                            Gucci Snakes
135                                        Wu-Tang Forever
144                               Primetime (feat. Miguel)
150                                            Rich Niggaz
242                                                Promise
529                                         Knockin' Boots
530                                  Thuggish Ruggish Bone
542                                              Oh My God
587                                         Patron Tequila
1357                                           Running Out
1387                                                  Zimt
1392                                Magnets - A-Trak Remix
1432                       The Water Dance (feat. Pitbull)
1531                                         Dance With Me
1700                         The Day You Stop Lookin' Back
1734                               I Co

Predicted songs
 22                                          Big Tymin
117                                       sobeautiful
147                       P.D.A. (We Just Don't Care)
242                                           Promise
331                      Run [I'm A Natural Disaster]
650                                  Get Buck In Here
1156                                          Lucifer
1203                           All The Way Up (Remix)
1384                                    Broken Arrows
1525           U And Dat (feat. T. Pain & Kandi Girl)
1592             19 Somethin' - Greatest Hits Version
2046                               Somewhere With You
2053                                          Payback
2125                                             This
2157                                             Love
2159                              Somewhere In My Car
2960                              Bad - 2012 Remaster
3193                                       Yesterdays
3210       

Predicted songs
 119                                               Muse
246                                             Get Up
578                             Dirt Off Your Shoulder
807                                       Slip N Slide
850                                        Left, Right
921                                            Tuesday
1141                                             Sorry
1168                                    Broadcast Live
1262     Bout Me - feat. Problem & Iamsu [Bonus Track]
1271                          Jealous - Bakermat Remix
1371                                      Stick Around
1775         Let It Go - Live From Spotify London/2015
2377          Take The Long Way Home - 2010 Remastered
2398                                   Learning To Fly
2603                        Little Miss Can't Be Wrong
2886                                   No Interruption
2998                                         The Metal
3003                         Freedom! '90 - Rema

Predicted songs
 124                  Boyfriend #2 - Explicit Album Version
135                                        Wu-Tang Forever
525                                         Me, Myself & I
721                                         Air Force Ones
751         Devil (feat. Busta Rhymes, B.o.B & Neon Hitch)
1138                                           I'm Goin In
1457                                      We Got The Party
1945                                           Dayum, Baby
2140                                            Neon Light
2158                                 Kick It In The Sticks
2310             Learning To Fly - 2011 Remastered Version
2408                                      Rock & Roll Band
2430                              Little Lies - Remastered
2481                           Waiting For A Girl Like You
2625                                              Spoonman
3060                                        Mr. Brownstone
3215                                   

Predicted songs
 45                                                     WOA
326                                               I. crawl
486                                     The World Is Yours
496                              Mind Playing Tricks on Me
591                                             Tambourine
719                                    Levels - Radio Edit
767                                  Writing's On The Wall
1193     I Think They Like Me - feat. Jermaine Dupri, D...
1233     Shoulder Lean (feat. T.I.) - Featuring T.I. Ex...
1367                                    Whole Lotta Lovin'
1435            Run Up (feat. PARTYNEXTDOOR & Nicki Minaj)
1733                                            My Old Man
1892                                           Kiss a Girl
1967                                        Goodnight Kiss
2202                                     Piece of My Heart
2518                                        I Will Survive
2562                          Shake Ya 

Predicted songs
 66                              Wat U Mean (Aye, Aye, Aye)
92                                                 Wife Er
671      Love Me Like You Do - From "Fifty Shades Of Grey"
715                                                 Switch
732                                             Girlfriend
736                                               Hangover
1222                          We Be Burnin' (Recognize It)
1267                                Let It Go - feat. Akon
1301                        Bad (feat. Vassy) - Radio Edit
1418                                       Rest Of My Life
1461                                      One In A Million
1541                                               We Rock
1579                                           Bait A Hook
1713                                      Let Me Down Easy
1924                                             Ex To See
2017                                                 Noise
2058                                   

Predicted songs
 28                                                Feedback
303                                             In One Ear
339                                    Steady, As She Goes
719                                    Levels - Radio Edit
1232                                              No Limit
1415                          Let Me Hold You (Turn Me On)
1457                                      We Got The Party
1507                                G.N.O. Girls Night Out
1596                                                Mayday
1866                                                 Human
2000                                     Good Lookin' Girl
2136                                     Bumpin' The Night
2140                                            Neon Light
2585                                         I Stand Alone
2611                                         Pure Massacre
3358                                            Big Amount
3382                                   

Predicted songs
 45                                              WOA
117                                     sobeautiful
242                                         Promise
350                                      Obstacle 1
495             Int'l Players Anthem (I Choose You)
587                                  Patron Tequila
591                                      Tambourine
677                                            Donk
715                                          Switch
763                                          Hollow
838                                   Body Language
1255                               Something Better
1288         Five More Hours - Deorro x Chris Brown
1302                           Heroes (we could be)
1383                                  Pure Grinding
1416                              Perfect Strangers
1432                The Water Dance (feat. Pitbull)
1435     Run Up (feat. PARTYNEXTDOOR & Nicki Minaj)
1551             Strut - From "The Cheetah Girl

Predicted songs
 124                  Boyfriend #2 - Explicit Album Version
146                                           Made to Love
584                  Return Of The Mack - C & J Street Mix
720                                              Beat Down
761                                                Perfect
842                                         Si Tú No Estás
1131                                               Mercy.1
1204     Lean Back (feat. Lil Jon, Eminem, Mase & Remy ...
1288                Five More Hours - Deorro x Chris Brown
1302                                  Heroes (we could be)
1321                                    Once In A Lifetime
1417                       Send It (feat. Rich Homie Quan)
1457                                      We Got The Party
1903                                       Time Well Spent
1945                                           Dayum, Baby
1956                                         She's Country
1993                                   

Predicted songs
 137                                             Love Faces
145                         LoveHate Thing - feat. Sam Dew
583         Can I Get A... - From The Rush Hour Soundtrack
715                                                 Switch
761                                                Perfect
791                                           Pretty Hurts
910      Let's Go (feat. Big D & Twista) - explicit alb...
1255                                      Something Better
1340     Feeling Right (Everything Is Nice) [feat. Popc...
1384                                         Broken Arrows
1695                                    Here Comes Goodbye
1890                                    First Time Feeling
2157                                                  Love
2159                                   Somewhere In My Car
2357                                   Something About You
2584                                               Runaway
2822                                   

Predicted songs
 137                                             Love Faces
260                                             I Want Her
791                                           Pretty Hurts
948                                          Calabria 2008
1084                                            Basketball
1232                                              No Limit
1424                                                 Alarm
1507                                G.N.O. Girls Night Out
1559     Only You-Bad Boy Remix (feat. The Notorious B....
1596                                                Mayday
1719               Sober Saturday Night (feat. Vince Gill)
1770                                      Used To Love You
1807                                              Been You
1866                                                 Human
1900                You Can't Stop Me (feat. Thomas Rhett)
2157                                                  Love
2159                                   

Predicted songs
 22                                               Big Tymin
82                                      Look Alive - Remix
109                                           Arch & Point
343                                        I Melt With You
578                                 Dirt Off Your Shoulder
842                                         Si Tú No Estás
871                                            Rock Bottom
921                                                Tuesday
1417                       Send It (feat. Rich Homie Quan)
1993                                                 Gonna
2035                                       She Get Me High
2198                                             Strangers
2320                                      Live And Let Die
2747                                         Wobble Wobble
2998                                             The Metal
3295                                              The Ride
3330                      Earned It (Fi

Predicted songs
 12                                      King Of Everything
98                            Pull Up (feat. Lil Uzi Vert)
165                                 Neighbors Know My Name
485            Chickenhead (feat. La Chat & Three 6 Mafia)
495                    Int'l Players Anthem (I Choose You)
716                                              Got Money
738                       Home - Fedde Le Grand Radio Edit
1045                             Bright Lights Bigger City
1104                                              Research
1273                      Church - Main Version - Explicit
1288                Five More Hours - Deorro x Chris Brown
1302                                  Heroes (we could be)
1375                        The Nights - Felix Jaehn Remix
1416                                     Perfect Strangers
1482               Dirt Off Your Shoulder / Lying From You
1551                    Strut - From "The Cheetah Girls 2"
1695                                   

Predicted songs
 19                          Bake Sale (feat. Travis Scott)
150                                            Rich Niggaz
334                                             Blue Jeans
530                                  Thuggish Ruggish Bone
713                            Aaron's Party (Come Get It)
795          Beautiful Liar - Main Version / Album Version
825                                            Don't Worry
881                    Mind Your Manners (feat. Icona Pop)
1194                                             White Tee
1285     Crazy Youngsters - From "Pitch Perfect 2" Soun...
1383                                         Pure Grinding
1387                                                  Zimt
1404                                 Heatwave (feat. Akon)
1418                                       Rest Of My Life
1551                    Strut - From "The Cheetah Girls 2"
1674                                                  Amen
1688                                   

Predicted songs
 148                                        Fantasy
155                                        So Gone
487                              Left, Right, Left
518                              Scenario - LP Mix
539                                  Ms. Fat Booty
592                             Often - Kygo Remix
1220                                Duffle Bag Boy
1249               Waves - Robin Schulz Radio Edit
1309                                      Coastin'
1322                             Let's Do It Again
1358                                       Save Me
1708                           Real Men Love Jesus
2055                                         Banjo
2058                                    We Were Us
2128                                        Voices
2310     Learning To Fly - 2011 Remastered Version
2326                   Over The Hills And Far Away
2362                      The Great Gig in the Sky
2430                      Little Lies - Remastered
2455          

### Measuring success

To measure success for the baseline model, we compute the mean of the R-precision scores across all the playlists in the test set. This returns us a value of 0.005, which implies that on average, out of all the withheld songs in the test playlists, we only manage to predict 0.8% of them using our model. This is low -  as a baseline model, it gives us a sense of the challenges that lie ahead in our predictions. 

In [45]:
avg_rp = np.mean(r_precision_scores)

print('Avg. R-Precision Score: ', avg_rp)

Avg. R-Precision Score:  0.008311438433374943


### Challenges and next steps

One of the significant challenges that we have identified could be the fact that the withheld songs in the test data set are not in our original training set. Hence, logically, the model trained on songs in the training set would not be able to return predicted songs that would match 1-to-1 with the withheld songs.

To validate this hypothesis, we ran a quick check on the intersection of songs between both sets, and as expected, only 1552 songs in the withheld list of songs, out of a total of 11074 songs, appear in the training set.

Going ahead, this will mean possibly expanding the training set to increase the overlap in songs between the two sets.

In [47]:
def intersection(lst1, lst2): 
    lst3 = [value for value in lst1 if value in lst2] 
    return lst3 
  
print (len(intersection(list(withheld.track_uri), list(train.index))))
print (len(list(withheld.track_uri)))

1552
11074
